Add support for ceph version luminous

AlexZzz opened the pull request:
This patch is not backward compatible with previous ceph versions.
It adds support for [this](https://github.com/ceph/ceph/commit/767502e26ac0cfff5f060fe2ff9eca88de31258d#diff-b6cb32f55d238418c321e793d17d0da3) commit.

octo commented on src/ceph.c:
```diff
@@ -913,27 +914,34 @@ static int node_handler_fetch_data(void *arg, const char *val,
+    } else if (vtmp->avgcount_exists == 1 ) {
       vtmp->avgcount_exists = -1;
+      // skip this step if no need in long run latency
+      if (!long_run_latency_avg) {
```
I don't understand the logic of `long_run_latency_avg` – whether it is set to true or false, the user will always get a latency average … Why?

AlexZzz commented:
Hi @octo! Thank you for the answer and i'm sorry for poor description of the request.
This patch fixes issue [#2343](https://github.com/collectd/collectd/pull/2464). We've faced with the same error after upgrade ceph to luminous and made this patch to fix the problem.

AlexZzz commented on src/ceph.c:
```diff
@@ -913,27 +914,34 @@ static int node_handler_fetch_data(void *arg, const char *val,
-      double sum, result;
-      sscanf(val, "%lf", &sum);
-
+    } else if (vtmp->avgcount_exists == 0) {
```
As I can see - ```ds_name``` has a long metric name, not just ```avgcount``` or ```sum```. So, I replaced ```avgcount_exists``` with readable enum.

octo replied:
That's still a state machine and as a result extremely brittle. It's sensitive to new metrics being added and depends on the order of metrics.

Can't you use the existing `has_suffix()` function to determine which metric it is?

AlexZzz replied:
I tried to find easy way to do it and did some tests.
Unfortunately, ```key``` and ```ds_name``` doesn't contain suffix we need. ```ceph_cb_number``` cut suffix for latency metrics. Do you think it should be fixed within this pull request?

octo replied:
Yes, I think this should be fixed in this PR: the reason we have to revisit this code for a new Ceph release it because it is so brittle. A good fix for this problem should ensure that the code works even if Ceph should decide to

*   add a median, 95th percentile or distribution of the latency, or
*   re-order the fields, e.g. "sum" first, "avgcount" second.

AlexZzz commented:
Hi @octo !

I've changed code to make ```node_handler_fetch_data()``` stateless. Also, added a warning if unknown latency metric caught and add ```cut_suffix()``` for cutting suffix.
This code was tested on Jewel and Luminous ceph. So, it is backward compatible with previous ceph versions - not just "works on luminous".
Let me know if it needs more changes or I missed something!

AlexZzz commented:
Hi, @octo ,
Thanks for help with this patch.
I did some changes and it looks better now. Tests are passed.

octo commented on src/ceph.c:
```diff
@@ -522,18 +517,24 @@ static size_t count_parts(char const *key) {
+    } else if (has_suffix(key_str, ".avgtime")) {
+      cut_suffix(tmp, tmp_size, key_str, ".avgtime");
+    } else {
+      sstrncpy(tmp, key_str, sizeof(tmp));
```
You're using `tmp_size` above and `sizeof(tmp)` here. Please unify this. Since `tmp_size` is not modified, it's more common to just use `sizeof(tmp)` over and over again.

octo commented on src/ceph.c:
```diff
@@ -522,18 +517,24 @@ static size_t count_parts(char const *key) {
-      sz = sizeof(tmp);
-    sstrncpy(tmp, key_str, sz);
+  /* Strip suffix if it is ".type" or one of latency metric suffix. */
+  if (count_parts(key_str) > 2) {
```
At your option:

```c
/* initialize tmp */
sstrncpy(tmp, key_str, sizeof(tmp));

char const *cut_suffixes[] = {".type", ".avgcount", ".sum", ".avgtime"};
for (size_t i = 0; i < STATIC_ARRAY_SIZE(cut_suffixes); i++) {
  if (has_suffix(key_str, cut_suffixes[i])) {
    cut_suffix(tmp, sizeof(tmp), key_str, cut_suffixes[i]);
    break;
  }
}
```

This has less repetition and makes it easier to see just which suffixes are being cut.

octo commented on src/ceph.c:
```diff
@@ -507,6 +485,23 @@ static _Bool has_suffix(char const *str, char const *suffix) {
   return 0;
 }
 
+static int cut_suffix(char new_str[], size_t new_str_len, char const *str,
```
1.  The most common name for the first two arguments are `buffer` and `buffer_size`.
2.  Please use a `char *` instead of a `char []`. (Both are correct and mean essentially the same thing, but the latter is very uncommon, at least within the C codebase).
3.  You only ever return the value zero and you never check the return value. Make this a `void` function.

```c
static void cut_suffix(char *buffer, size_t buffer_size, ...)
```

octo commented on src/ceph.c:
```diff
@@ -907,34 +908,41 @@ static int node_handler_fetch_data(void *arg, const char *val,
       uv.gauge = result;
-      vtmp->avgcount_exists = -1;
       vtmp->latency_index = (vtmp->latency_index + 1);
+    } else if (has_suffix(key, ".avgtime")) {
```
Please add a comment here along the lines of:

> The "avgtime" metric reports ("sum" / "avgcount"), i.e. the average time per request since the start of the Ceph daemon. Report this only when the user has configured "long running average". Otherwise, use the *rate* of "sum" and "avgcount" to calculate the current latency.

AlexZzz commented:
Hi, @octo, thank you for review!
I did requested changes:)

octo commented:
Thanks @AlexZzz!
